File: src/chain-observers/providers/exchange-transfer-provider.ts
import type { asset } from "@hiveio/wax";
import { WorkerBeeIterable } from "../../types/iterator";
import { Exchange, isExchange } from "../../utils/known-exchanges";
import { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import { OperationClassifier, IOperationTransactionPair } from "../classifiers/operation-classifier";
import { TProviderEvaluationContext } from "../factories/data-evaluation-context";
import { ProviderBase } from "./provider-base";

export interface IExchangeTransferMetadata {
  from: string;
  to: string;
  amount: asset;
  exchange: Exchange;
}

export interface IExchangeTransferProviderData {
  exchangeTransferOperations: WorkerBeeIterable<IOperationTransactionPair<IExchangeTransferMetadata>>;
};

/**
 * Note: On escrow_transfer operation, hbd_amount is used as the general-purpose amount field.
 * This is because the escrow_transfer operation is used for both HIVE and HBD transfers.
 * If you want to extract the HIVE amount, you should extract it directly from the provided operations within transaction.
 */
export class ExchangeTransferProvider extends ProviderBase<{}, IExchangeTransferProviderData> {
  public usedContexts(): Array<TRegisterEvaluationContext> {
    return [
      OperationClassifier
    ]
  }

  public get baseStructure(): IExchangeTransferProviderData {
    return {
      exchangeTransferOperations: new WorkerBeeIterable([])
    };
  }

  public async provide(data: TProviderEvaluationContext): Promise<IExchangeTransferProviderData> {
    const operations = await data.get(OperationClassifier);

    const transfer = operations.operationsPerType["transfer_operation"];
    const fromSavings = operations.operationsPerType["transfer_from_savings_operation"];
    const escrow = operations.operationsPerType["escrow_transfer_operation"];
    const recurrent = operations.operationsPerType["recurrent_transfer_operation"];

    const exchangeTransfers: IOperationTransactionPair<IExchangeTransferMetadata>[] = [];

    if(transfer)
      for(const op of transfer) {
        const fromExchange = isExchange(op.operation.from);
        const toExchange = isExchange(op.operation.to);

        if (fromExchange || toExchange)
          exchangeTransfers.push({
            operation: {
              from: op.operation.from,
              to: op.operation.to,
              amount: op.operation.amount!,
              exchange: (fromExchange || toExchange) as Exchange
            },
            transaction: op.transaction
          });
      }

    if(fromSavings)
      for(const op of fromSavings) {
        const fromExchange = isExchange(op.operation.from);
        const toExchange = isExchange(op.operation.to);

        if (fromExchange || toExchange)
          exchangeTransfers.push({
            operation: {
              from: op.operation.from,
              to: op.operation.to,
              amount: op.operation.amount!,
              exchange: (fromExchange || toExchange) as Exchange
            },
            transaction: op.transaction
          });
      }

    if(escrow)
      for(const op of escrow) {
        const fromExchange = isExchange(op.operation.from);
        const toExchange = isExchange(op.operation.to);

        if (fromExchange || toExchange)
          exchangeTransfers.push({
            operation: {
              from: op.operation.from,
              to: op.operation.to,
              amount: op.operation.hbd_amount!,
              exchange: (fromExchange || toExchange) as Exchange
            },
            transaction: op.transaction
          });
      }

    if(recurrent)
      for(const op of recurrent) {
        const fromExchange = isExchange(op.operation.from);
        const toExchange = isExchange(op.operation.to);

        if (fromExchange || toExchange)
          exchangeTransfers.push({
            operation: {
              from: op.operation.from,
              to: op.operation.to,
              amount: op.operation.amount!,
              exchange: (fromExchange || toExchange) as Exchange
            },
            transaction: op.transaction
          });
      }

    const result = this.baseStructure;

    result.exchangeTransferOperations = new WorkerBeeIterable(exchangeTransfers);

    return result;
  }
}

File: python/workerbee/chain_observers/providers/exchange_transfer_provider.py
"""ExchangeTransferProvider — provides transfer operations involving known exchanges.

Mirrors src/chain-observers/providers/exchange-transfer-provider.ts.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
from ..classifiers.operation_classifier import OperationClassifier
from ..utils import is_exchange
from ._transfer_helpers import TRANSFER_OP_TYPES, get_transfer_amount
from .provider_base import ProviderBase

if TYPE_CHECKING:
    from ..factories.data_evaluation_context import DataEvaluationContext
    from ..payloads import ExchangeTransferPair, ExchangeTransferPayload


class ExchangeTransferProvider(ProviderBase):
    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        return [OperationClassifier]

    async def provide(self, data: DataEvaluationContext) -> ExchangeTransferPayload:
        operations = await data.get(OperationClassifier)
        ops_per_type = operations["operations_per_type"]

        exchange_transfers: list[ExchangeTransferPair] = []

        for op_type in TRANSFER_OP_TYPES:
            ops = ops_per_type.get(op_type)
            if not ops:
                continue

            for op in ops:
                operation = op["operation"]
                from_exchange = is_exchange(str(operation.get("from", "")))
                to_exchange = is_exchange(str(operation.get("to", "")))
                if not from_exchange and not to_exchange:
                    continue

                amount = get_transfer_amount(op, op_type)
                exchange_transfers.append(
                    cast(
                        "ExchangeTransferPair",
                        {
                            "operation": {
                                "from": operation.get("from", ""),
                                "to": operation.get("to", ""),
                                "amount": amount,
                                "exchange": from_exchange or to_exchange,
                            },
                            "transaction": op["transaction"],
                        },
                    )
                )

        return {"exchange_transfer_operations": exchange_transfers}
